From 08bf6b4a281ef4308119dccbba4e86cf28b505d2 Mon Sep 17 00:00:00 2001
From: Sumit Bose <sbose@redhat.com>
Date: Fri, 4 Nov 2016 17:13:30 +0100
Subject: [PATCH 71/79] sdap_extend_map: make sure memory can be freed

If there is an error after calling talloc_realloc() the caller cannot
free the memory properly because neither src_map nor _map were pointing
to a valid memory location. With this patch _map will always point to
the current valid location so that it can always be used with
talloc_free().

Reviewed-by: Petr Cech <pcech@redhat.com>
---
 src/providers/ldap/sdap.c |  4 ++--
 src/providers/ldap/sdap.h | 21 +++++++++++++++++++++
 2 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/src/providers/ldap/sdap.c b/src/providers/ldap/sdap.c
index bfb7fc6d2a38debf56acae18b8e7eb7a08ccbd1b..342667aae6cc0b6c3df8c76f11eb2445e6bad0be 100644
--- a/src/providers/ldap/sdap.c
+++ b/src/providers/ldap/sdap.c
@@ -162,9 +162,9 @@ int sdap_extend_map(TALLOC_CTX *memctx,
     char *sysdb_attr;
     errno_t ret;
 
+    *_map = src_map;
     if (extra_attrs == NULL) {
         DEBUG(SSSDBG_FUNC_DATA, "No extra attributes\n");
-        *_map = src_map;
         *_new_size = num_entries;
         return EOK;
     }
@@ -177,6 +177,7 @@ int sdap_extend_map(TALLOC_CTX *memctx,
     if (map == NULL) {
         return ENOMEM;
     }
+    *_map = map;
 
     for (i = 0; *extra_attrs != NULL; extra_attrs++) {
         ret = split_extra_attr(map, *extra_attrs, &sysdb_attr, &ldap_attr);
@@ -221,7 +222,6 @@ int sdap_extend_map(TALLOC_CTX *memctx,
     /* Sentinel */
     memset(&map[num_entries+nextra], 0, sizeof(struct sdap_attr_map));
 
-    *_map = map;
     *_new_size = num_entries + nextra;
     return EOK;
 }
diff --git a/src/providers/ldap/sdap.h b/src/providers/ldap/sdap.h
index 6d4543ed48ce19f82252d34b6d0833a546a81bb9..6079a8bf62d0bdf23c8d462dc0f19c705e391a6e 100644
--- a/src/providers/ldap/sdap.h
+++ b/src/providers/ldap/sdap.h
@@ -512,6 +512,27 @@ int sdap_copy_map(TALLOC_CTX *memctx,
                  int num_entries,
                  struct sdap_attr_map **_map);
 
+/**
+ * @brief Add attributes to a map
+ *
+ * sdap_extend_map() will call talloc_realloc() on the second argument so the
+ * original storage location might change. The return value _map will always
+ * contain the current memory location which can be used with talloc_free()
+ * even if there is an error.
+ *
+ * @param[in] memctx      Talloc memory context
+ * @param[in] src_map     Original map, should not be accessed anymore
+ * @param[in] num_entries Number of entries in the original map
+ * @param[in] extra_attrs NULL-terminated array of extra attribute pairs
+ *                        sysdb_attr:ldap_attr
+ * @param[out] _map       New map
+ * @param[out] _new_size  Number of entries in the new map
+ *
+ * @return
+ *  - EOK                 success
+ *  - ENOMEM              memory allocation failed
+ *  - ERR_DUP_EXTRA_ATTR  sysdb attribute is already used
+ */
 int sdap_extend_map(TALLOC_CTX *memctx,
                     struct sdap_attr_map *src_map,
                     size_t num_entries,
-- 
2.9.3

